File: sim_eval.js
/* sim_eval.js — JS port of sim_eval.py (Option 1).
 *
 * Faithful 1:1 port of the DeltaV structured-text evaluator: tokenizer +
 * recursive-descent parser + tree-walking evaluator. Same grammar, same
 * coercion rules, so a walk driven by this engine reproduces the Python walk.
 *
 * Parity notes (where JS and Python differ, we force Python semantics):
 *  - truthiness: '' / 0 / false / null => falsy; everything else truthy.
 *  - numeric coercion: booleans -> 0/1 for arithmetic & numeric compare.
 *  - '=' is EQUALITY (not assignment) inside expressions; ':=' is assignment.
 *  - string vs number compare: if either side is a string, compare as strings
 *    (matches Python's branch); else compare numerically.
 *  - unknown refs read as 0 (falsy), exactly like store.get(key, 0).
 *
 * Exposed: tokenize, Parser, Evaluator, evalTransition, runActions, makeRef.
 * No DOM, no globals — safe to node --check and unit-test standalone.
 */
(function (root) {
  'use strict';

  // ── value model ──────────────────────────────────────────────────────────
  function Ref(key) { this.key = key; }
  function makeRef(key) { return new Ref(key); }
  function isRef(v) { return v instanceof Ref; }

  function truthy(v) {
    if (typeof v === 'boolean') return v;
    if (typeof v === 'number') return v !== 0;
    if (typeof v === 'string') return v !== '';
    return v !== null && v !== undefined;
  }

  // Coerce booleans to 0/1 so 'X = 1' and 'X = TRUE' both work.
  function num(v) {
    if (typeof v === 'boolean') return v ? 1 : 0;
    return v;
  }

  function asStr(v) {
    if (typeof v === 'boolean') return v ? 'True' : 'False';   // match Python repr
    if (typeof v === 'number' && Number.isInteger(v)) return String(v);
    return String(v);
  }

  // ── tokenizer ────────────────────────────────────────────────────────────
  // Mirrors the Python verbose regex, evaluated left-to-right with sticky flags.
  const KEYWORDS = new Set(
    ['AND', 'OR', 'NOT', 'IF', 'THEN', 'ELSE', 'END_IF', 'ENDIF', 'TRUE', 'FALSE']);

  // Order matters: longer operators before shorter prefixes.
  const SUBPATTERNS = [
    ['ws',   /\s+/y],
    ['ref',  /'(?:[^']|'')*'/y],            // 'a/b.CV' or 'SET:MEMBER'
    ['str',  /"[^"]*"/y],                    // "text"
    ['num',  /\d+\.\d+|\d+/y],
    ['op',   /:=|<=|>=|<>|!=|=|<|>|\+|\-|\*|\//y],
    ['punc', /[()]|;/y],
    ['word', /[A-Za-z_][A-Za-z0-9_]*/y],
  ];

  function tokenize(srcIn) {
    // strip (* ... *) comments, then un-double FHX-escaped quotes once.
    let src = srcIn.replace(/\(\*[\s\S]*?\*\)/g, '');
    src = src.split('""').join('"');
    const toks = [];
    let i = 0;
    const n = src.length;
    while (i < n) {
      let matched = false;
      for (let s = 0; s < SUBPATTERNS.length; s++) {
        const kind = SUBPATTERNS[s][0];
        const re = SUBPATTERNS[s][1];
        re.lastIndex = i;
        const m = re.exec(src);
        if (m && m.index === i) {
          matched = true;
          i = re.lastIndex;
          const val = m[0];
          if (kind === 'ws') { /* skip */ }
          else if (kind === 'ref') {
            toks.push(['ref', val.slice(1, -1).split("''").join("'")]);
          } else if (kind === 'str') {
            toks.push(['str', val.slice(1, -1)]);
          } else if (kind === 'num') {
            toks.push(['num', val.indexOf('.') >= 0 ? parseFloat(val) : parseInt(val, 10)]);
          } else if (kind === 'word') {
            const up = val.toUpperCase();
            toks.push([KEYWORDS.has(up) ? up : 'word', val]);
          } else { // op / punc
            const t = val.trim();
            toks.push([t, t]);
          }
          break;
        }
      }
      if (!matched) {
        throw new SyntaxError('cannot tokenize at: ' + JSON.stringify(src.slice(i, i + 30)));
      }
    }
    toks.push(['eof', null]);
    return toks;
  }

  // ── parser ───────────────────────────────────────────────────────────────
  function Parser(toks, namedSets) {
    this.t = toks;
    this.p = 0;
    this.sets = namedSets || {};
  }
  Parser.prototype._peek = function () { return this.t[this.p]; };
  Parser.prototype._next = function () { return this.t[this.p++]; };
  Parser.prototype._expect = function (kind) {
    const tok = this._next();
    if (tok[0] !== kind) throw new SyntaxError('expected ' + kind + ', got ' + JSON.stringify(tok));
    return tok;
  };

  Parser.prototype.parseExpr = function () { return this._or(); };
  Parser.prototype._or = function () {
    let node = this._and();
    while (this._peek()[0] === 'OR') { this._next(); node = ['or', node, this._and()]; }
    return node;
  };
  Parser.prototype._and = function () {
    let node = this._not();
    while (this._peek()[0] === 'AND') { this._next(); node = ['and', node, this._not()]; }
    return node;
  };
  Parser.prototype._not = function () {
    if (this._peek()[0] === 'NOT') { this._next(); return ['not', this._not()]; }
    return this._cmp();
  };
  Parser.prototype._cmp = function () {
    const node = this._add();
    const k = this._peek()[0];
    if (k === '=' || k === '!=' || k === '<>' || k === '<' || k === '>' || k === '<=' || k === '>=') {
      const op = this._next()[0];
      const rhs = this._add();
      return ['cmp', op === '<>' ? '!=' : op, node, rhs];
    }
    return node;
  };
  Parser.prototype._add = function () {
    let node = this._mul();
    while (this._peek()[0] === '+' || this._peek()[0] === '-') {
      const op = this._next()[0];
      node = ['bin', op, node, this._mul()];
    }
    return node;
  };
  Parser.prototype._mul = function () {
    let node = this._primary();
    while (this._peek()[0] === '*' || this._peek()[0] === '/') {
      const op = this._next()[0];
      node = ['bin', op, node, this._primary()];
    }
    return node;
  };
  Parser.prototype._primary = function () {
    const tok = this._peek();
    const kind = tok[0], val = tok[1];
    if (kind === '(') { this._next(); const node = this.parseExpr(); this._expect(')'); return node; }
    if (kind === 'num') { this._next(); return ['lit', val]; }
    if (kind === 'str') { this._next(); return ['lit', val]; }
    if (kind === 'TRUE') { this._next(); return ['lit', true]; }
    if (kind === 'FALSE') { this._next(); return ['lit', false]; }
    if (kind === 'ref') { this._next(); return this._refOrMember(val); }
    if (kind === 'word') { this._next(); return ['ref', val]; }
    throw new SyntaxError('unexpected token ' + JSON.stringify(this._peek()));
  };
  Parser.prototype._refOrMember = function (raw) {
    if (raw.indexOf(':') >= 0) {
      const idx = raw.indexOf(':');
      const setname = raw.slice(0, idx), member = raw.slice(idx + 1);
      const members = this.sets[setname];
      if (members !== undefined && members !== null) {
        return ['lit', (member in members) ? members[member] : raw];
      }
      return ['lit', raw];
    }
    return ['ref', new Ref(raw)];
  };

  // statements
  Parser.prototype.parseStmts = function () {
    const stmts = [];
    while (true) {
      const k = this._peek()[0];
      if (k === 'eof' || k === 'END_IF' || k === 'ENDIF' || k === 'ELSE') break;
      if (k === ';') { this._next(); continue; }
      stmts.push(this._stmt());
      if (this._peek()[0] === ';') this._next();
    }
    return ['block', stmts];
  };
  Parser.prototype._stmt = function () {
    if (this._peek()[0] === 'IF') {
      this._next();
      const cond = this.parseExpr();
      this._expect('THEN');
      const then = this.parseStmts();
      let els = null;
      if (this._peek()[0] === 'ELSE') { this._next(); els = this.parseStmts(); }
      const k = this._peek()[0];
      if (k === 'END_IF' || k === 'ENDIF') this._next();
      else this._expect('END_IF');
      return ['if', cond, then, els];
    }
    const tok = this._next();
    if (tok[0] !== 'ref') throw new SyntaxError('assignment must start with a ref, got ' + JSON.stringify(tok));
    this._expect(':=');
    return ['assign', new Ref(tok[1]), this.parseExpr()];
  };

  // ── evaluator ──────────────────────────────────────────────────────────────
  function Evaluator(store, namedSets) {
    this.store = store;
    this.sets = namedSets || {};
  }
  Evaluator.prototype._read = function (ref) {
    const v = this.store[ref.key];
    return v === undefined ? 0 : v;
  };
  Evaluator.prototype.eval = function (node) {
    const op = node[0];
    if (op === 'lit') return node[1];
    if (op === 'ref') {
      const r = node[1];
      if (isRef(r)) return this._read(r);
      const v = this.store[r];
      return v === undefined ? 0 : v;
    }
    if (op === 'not') return !truthy(this.eval(node[1]));
    if (op === 'bin') {
      const o = node[1];
      const l = this.eval(node[2]), r = this.eval(node[3]);
      if (o === '+') {
        if (typeof l === 'string' || typeof r === 'string') return asStr(l) + asStr(r);
        return num(l) + num(r);
      }
      if (o === '-') return num(l) - num(r);
      if (o === '*') return num(l) * num(r);
      if (o === '/') return num(r) ? num(l) / num(r) : 0;
    }
    if (op === 'and') return truthy(this.eval(node[1])) && truthy(this.eval(node[2]));
    if (op === 'or') return truthy(this.eval(node[1])) || truthy(this.eval(node[2]));
    if (op === 'cmp') {
      const c = node[1];
      let l = this.eval(node[2]), r = this.eval(node[3]);
      let ll, rr;
      if (typeof l === 'string' || typeof r === 'string') { ll = l; rr = r; }
      else { ll = num(l); rr = num(r); }
      if (c === '=') return ll === rr;
      if (c === '!=') return ll !== rr;
      if (c === '<') return ll < rr;
      if (c === '>') return ll > rr;
      if (c === '<=') return ll <= rr;
      if (c === '>=') return ll >= rr;
    }
    throw new Error('bad expr node ' + JSON.stringify(node));
  };
  Evaluator.prototype.exec = function (node) {
    const op = node[0];
    if (op === 'block') { for (let i = 0; i < node[1].length; i++) this.exec(node[1][i]); }
    else if (op === 'assign') { this.store[node[1].key] = this.eval(node[2]); }
    else if (op === 'if') {
      if (truthy(this.eval(node[1]))) this.exec(node[2]);
      else if (node[3] !== null) this.exec(node[3]);
    } else throw new Error('bad stmt node ' + JSON.stringify(node));
  };

  // ── public API ───────────────────────────────────────────────────────────
  function evalTransition(exprSrc, store, namedSets) {
    let src = (exprSrc || '').trim();
    while (src.endsWith(';')) src = src.slice(0, -1).trim();
    if (!src) return true;
    const toks = tokenize(src);
    const ast = new Parser(toks, namedSets).parseExpr();
    return truthy(new Evaluator(store, namedSets).eval(ast));
  }

  function runActions(actionSrc, store, namedSets) {
    if (!actionSrc || !actionSrc.trim()) return;
    const toks = tokenize(actionSrc);
    const ast = new Parser(toks, namedSets).parseStmts();
    new Evaluator(store, namedSets).exec(ast);
  }

  const API = {
    tokenize: tokenize, Parser: Parser, Evaluator: Evaluator,
    evalTransition: evalTransition, runActions: runActions,
    makeRef: makeRef, Ref: Ref, _truthy: truthy,
  };

  if (typeof module !== 'undefined' && module.exports) module.exports = API;
  else root.SimEval = API;
})(typeof window !== 'undefined' ? window : this);

File: sim_eval.py
"""
sim_eval.py  —  SIMULATOR SPIKE (task 2): a minimal evaluator for the DeltaV
structured-text subset used by phase SFC transitions and actions.

This is a spike: it imports nothing from the explorer and touches nothing in it.
Goal: prove the real transition/action expressions from a phase RUN sequence can
be tokenized, parsed, and evaluated against a flat variable store.

Grammar covered (derived from the actual CIP WASH RUN sequence):
  transitions (boolean):
      ref/literal compared with = != <> < > <= >=, combined with AND/OR/NOT,
      parenthesised, plus bare-boolean (truthy) terms.
  actions (statements):
      <ref> := <expr> ;            assignment
      IF (<expr>) THEN <stmts> [ELSE <stmts>] END_IF ;
      ;-separated, multi-line statement sequences.

References resolve to a flat dict by their raw text (quotes stripped); within one
phase the reference strings are internally consistent, so no scope resolution is
needed for the spike. Named-set members ('SET:MEMBER') resolve to their integer
value when SET is known. Unknown variables read as 0 (falsy).
"""

import re


# ───────────────────────────── value model ──────────────────────────────────
class Ref:
    __slots__ = ('key',)

    def __init__(self, key):
        self.key = key

    def __repr__(self):
        return f'Ref({self.key!r})'


def _truthy(v):
    if isinstance(v, bool):
        return v
    if isinstance(v, (int, float)):
        return v != 0
    if isinstance(v, str):
        return v != ''
    return v is not None


def _num(v):
    """Coerce booleans to 0/1 so 'X = 1' and 'X = TRUE' both work."""
    if isinstance(v, bool):
        return 1 if v else 0
    return v


def _as_str(v):
    if isinstance(v, bool):
        return 'True' if v else 'False'
    if isinstance(v, float) and v.is_integer():
        return str(int(v))
    return str(v)


def _numf(v):
    """Numeric coercion for math builtins."""
    v = _num(v)
    try:
        return float(v)
    except (TypeError, ValueError):
        return 0.0


def _call_builtin(fn, args):
    """DeltaV structured-text builtins (mostly used in message construction).
    Unknown functions pass through their first arg so a message still renders."""
    import math
    if fn == 'round':
        return int(round(_numf(args[0])))
    if fn in ('trunc', 'int'):
        return int(_numf(args[0]))
    if fn == 'abs':
        return abs(_numf(args[0]))
    if fn == 'float':
        return _numf(args[0])
    if fn == 'sqrt':
        return math.sqrt(_numf(args[0]))
    if fn == 'min':
        r = min(_numf(a) for a in args)
        return int(r) if r == int(r) else r
    if fn == 'max':
        r = max(_numf(a) for a in args)
        return int(r) if r == int(r) else r
    if fn in ('time', 'time_to_str'):
        return '<time>'
    if fn == 'date_to_str':
        return '<date>'
    if fn == 'str':
        return _as_str(args[0]) if args else ''
    if fn == 'str_to_int':
        try:
            return int(_as_str(args[0]))
        except (ValueError, TypeError):
            return 0
    if fn == 'str_to_float':
        try:
            return float(_as_str(args[0]))
        except (ValueError, TypeError):
            return 0.0
    if fn == 'len':
        return len(_as_str(args[0])) if args else 0
    return args[0] if args else 0   # unknown fn: pass through


# ───────────────────────────── tokenizer ────────────────────────────────────
_TOKEN_RE = re.compile(r"""
    (?P<ws>\s+)
  | (?P<ref>'(?:[^']|'')*')                 # 'a/b.CV'  or  'SET:MEMBER'
  | (?P<str>"[^"]*")                         # "text"  (quotes already un-doubled)
  | (?P<denum>\$[A-Za-z_][A-Za-z0-9_]*:[^)'"\s,;]+)   # $time_format:Local
  | (?P<num>\d+\.\d+|\d+)
  | (?P<op>:=|<=|>=|<>|!=|=|<|>|\+|\-|\*|/)
  | (?P<punc>[(),] | ;)
  | (?P<word>[A-Za-z_][A-Za-z0-9_]*)
""", re.VERBOSE)

_KEYWORDS = {'AND', 'OR', 'NOT', 'IF', 'THEN', 'ELSE', 'END_IF', 'ENDIF', 'TRUE', 'FALSE'}


def tokenize(src):
    # strip (* ... *) block comments and `rem ...` line comments, then un-double
    # the FHX-escaped quotes once.
    src = re.sub(r'\(\*.*?\*\)', '', src, flags=re.DOTALL)
    src = re.sub(r'(^|[;\r\n])[ \t]*rem\b[^\r\n]*', r'\1', src, flags=re.IGNORECASE)
    src = src.replace('""', '"')
    toks, i, n = [], 0, len(src)
    while i < n:
        m = _TOKEN_RE.match(src, i)
        if not m:
            raise SyntaxError(f'cannot tokenize at: {src[i:i+30]!r}')
        i = m.end()
        kind = m.lastgroup
        if kind == 'ws':
            continue
        val = m.group()
        if kind == 'ref':
            toks.append(('ref', val[1:-1].replace("''", "'")))
        elif kind == 'str':
            toks.append(('str', val[1:-1]))
        elif kind == 'denum':
            toks.append(('str', val))   # $time_format:Local -> literal string
        elif kind == 'num':
            toks.append(('num', float(val) if '.' in val else int(val)))
        elif kind == 'word':
            up = val.upper()
            toks.append((up if up in _KEYWORDS else 'word', val))
        else:  # op / punc
            toks.append((val.strip(), val.strip()))
    toks.append(('eof', None))
    return toks


# ───────────────────────────── parser ───────────────────────────────────────
class Parser:
    def __init__(self, toks, named_sets=None):
        self.t = toks
        self.p = 0
        self.sets = named_sets or {}

    def _peek(self):
        return self.t[self.p]

    def _next(self):
        tok = self.t[self.p]
        self.p += 1
        return tok

    def _expect(self, kind):
        tok = self._next()
        if tok[0] != kind:
            raise SyntaxError(f'expected {kind}, got {tok}')
        return tok

    # -- expressions --
    def parse_expr(self):
        return self._or()

    def _or(self):
        node = self._and()
        while self._peek()[0] == 'OR':
            self._next()
            node = ('or', node, self._and())
        return node

    def _and(self):
        node = self._not()
        while self._peek()[0] == 'AND':
            self._next()
            node = ('and', node, self._not())
        return node

    def _not(self):
        if self._peek()[0] == 'NOT':
            self._next()
            return ('not', self._not())
        return self._cmp()

    def _cmp(self):
        node = self._add()
        if self._peek()[0] in ('=', '!=', '<>', '<', '>', '<=', '>='):
            op = self._next()[0]
            rhs = self._add()
            return ('cmp', '!=' if op == '<>' else op, node, rhs)
        return node

    def _add(self):
        node = self._mul()
        while self._peek()[0] in ('+', '-'):
            op = self._next()[0]
            node = ('bin', op, node, self._mul())
        return node

    def _mul(self):
        node = self._primary()
        while self._peek()[0] in ('*', '/'):
            op = self._next()[0]
            node = ('bin', op, node, self._primary())
        return node

    def _primary(self):
        kind, val = self._peek()
        if kind == '(':
            self._next()
            node = self.parse_expr()
            self._expect(')')
            return node
        if kind == 'num':
            self._next(); return ('lit', val)
        if kind == 'str':
            self._next(); return ('lit', val)
        if kind == 'TRUE':
            self._next(); return ('lit', True)
        if kind == 'FALSE':
            self._next(); return ('lit', False)
        if kind == 'ref':
            self._next()
            return self._ref_or_member(val)
        if kind == 'word':
            self._next()
            # function call?  word( args )
            if self._peek()[0] == '(':
                self._next()
                args = []
                if self._peek()[0] != ')':
                    args.append(self.parse_expr())
                    while self._peek()[0] == ',':
                        self._next(); args.append(self.parse_expr())
                self._expect(')')
                return ('call', val.lower(), args)
            return ('ref', val)
        raise SyntaxError(f'unexpected token {self._peek()}')

    def _ref_or_member(self, raw):
        # 'SET:MEMBER' with a known set -> integer literal; else a variable ref.
        if ':' in raw:
            setname, _, member = raw.partition(':')
            members = self.sets.get(setname)
            if members is not None:
                return ('lit', members.get(member, raw))
            # unknown set: keep as opaque string literal so comparisons still work
            return ('lit', raw)
        return ('ref', Ref(raw))

    # -- statements --
    def parse_stmts(self):
        stmts = []
        while self._peek()[0] not in ('eof', 'END_IF', 'ENDIF', 'ELSE'):
            if self._peek()[0] == ';':
                self._next(); continue
            stmts.append(self._stmt())
            if self._peek()[0] == ';':
                self._next()
        return ('block', stmts)

    def _stmt(self):
        if self._peek()[0] == 'IF':
            self._next()
            cond = self.parse_expr()
            self._expect('THEN')
            then = self.parse_stmts()
            els = None
            if self._peek()[0] == 'ELSE':
                self._next()
                els = self.parse_stmts()
            if self._peek()[0] in ('END_IF', 'ENDIF'):
                self._next()
            else:
                self._expect('END_IF')
            return ('if', cond, then, els)
        # assignment:  ref := expr  (ref is normally '^/...'; also allow a bare
        # word for DeltaV local/temporary variables like COUNT := 0)
        kind, val = self._next()
        if kind not in ('ref', 'word'):
            raise SyntaxError(f'assignment must start with a ref, got {(kind, val)}')
        self._expect(':=')
        return ('assign', Ref(val), self.parse_expr())


# ───────────────────────────── evaluator ────────────────────────────────────
class Evaluator:
    def __init__(self, store, named_sets=None):
        self.store = store          # flat dict: key -> value
        self.sets = named_sets or {}

    def _read(self, ref):
        return self.store.get(ref.key, 0)

    def eval(self, node):
        op = node[0]
        if op == 'lit':
            return node[1]
        if op == 'ref':
            r = node[1]
            return self._read(r) if isinstance(r, Ref) else self.store.get(r, 0)
        if op == 'not':
            return not _truthy(self.eval(node[1]))
        if op == 'call':
            return _call_builtin(node[1], [self.eval(a) for a in node[2]])
        if op == 'bin':
            _, o, a, b = node
            l, r = self.eval(a), self.eval(b)
            if o == '+':
                if isinstance(l, str) or isinstance(r, str):
                    return _as_str(l) + _as_str(r)        # string concat
                return _num(l) + _num(r)
            if o == '-':
                return _num(l) - _num(r)
            if o == '*':
                return _num(l) * _num(r)
            if o == '/':
                return _num(l) / _num(r) if _num(r) else 0
        if op == 'and':
            return _truthy(self.eval(node[1])) and _truthy(self.eval(node[2]))
        if op == 'or':
            return _truthy(self.eval(node[1])) or _truthy(self.eval(node[2]))
        if op == 'cmp':
            _, c, a, b = node
            l, r = self.eval(a), self.eval(b)
            if isinstance(l, str) or isinstance(r, str):
                ll, rr = l, r          # string compare as-is
            else:
                ll, rr = _num(l), _num(r)
            if c == '=':
                return ll == rr
            if c == '!=':
                return ll != rr
            if c == '<':
                return ll < rr
            if c == '>':
                return ll > rr
            if c == '<=':
                return ll <= rr
            if c == '>=':
                return ll >= rr
        raise ValueError(f'bad expr node {node}')

    def exec(self, node):
        op = node[0]
        if op == 'block':
            for s in node[1]:
                self.exec(s)
        elif op == 'assign':
            self.store[node[1].key] = self.eval(node[2])
        elif op == 'if':
            if _truthy(self.eval(node[1])):
                self.exec(node[2])
            elif node[3] is not None:
                self.exec(node[3])
        else:
            raise ValueError(f'bad stmt node {node}')


# ───────────────────────────── public API ───────────────────────────────────
def eval_transition(expr_src, store, named_sets=None):
    """Evaluate a transition expression -> bool. Trailing ';' is tolerated."""
    src = expr_src.strip().rstrip(';')
    if not src:
        return True
    toks = tokenize(src)
    ast = Parser(toks, named_sets).parse_expr()
    return _truthy(Evaluator(store, named_sets).eval(ast))


def run_actions(action_src, store, named_sets=None):
    """Execute an action body (statement sequence) against the store in place."""
    if not action_src or not action_src.strip():
        return
    toks = tokenize(action_src)
    ast = Parser(toks, named_sets).parse_stmts()
    Evaluator(store, named_sets).exec(ast)
